provider/cmd/crd2pulumi/generate.go

```diff
@@ -0,0 +1,209 @@
+	}
+
+	versions := map[string]map[string]interface{}{}
+
+	validation, foundValidation, _ := unstruct.NestedMap(crd.Object, "spec", "validation", "openAPIV3Schema")
+	if foundValidation { // If present, use the top-level schema to validate all versions
+		versionMaps, _, _ := NestedMapSlice(crd.Object, "spec", "versions")
+		for _, version := range versionMaps {
+			name, _, _ := unstruct.NestedString(version, "name")
+			versions[name] = validation
+		}
+	} else { // Otherwise use per-version schemas to validate each version
+		versionMaps, _, _ := NestedMapSlice(crd.Object, "spec", "versions")
+		for _, version := range versionMaps {
+			name, _, _ := unstruct.NestedString(version, "name")
+			schema, _, _ := unstruct.NestedMap(version, "schema", "openAPIV3Schema")
+			versions[name] = schema
+		}
+	}
+
+	kind, _, _ := unstruct.NestedString(crd.Object, "spec", "names", "kind")
+	plural, _, _ := unstruct.NestedString(crd.Object, "spec", "names", "plural")
+	group, _, _ := unstruct.NestedString(crd.Object, "spec", "group")
+
+	if outputPath == "" {
+		defaultOutputPath := func(yamlPath, plural, language string) string {
+			var extension string
+			switch language {
+			case NodeJS:
+				extension = "ts"
+			case DotNet:
+				extension = "cs"
+			case Python:
+				extension = "py"
+			case Go:
+				extension = "go"
+			}
+			outputFileName := plural + "." + extension
+			return path.Join(filepath.Dir(yamlPath), outputFileName)
+		}
+		outputPath = defaultOutputPath(yamlPath, plural, language)
+	}
+
+	customResourceGenerator := CustomResourceGenerator{
+		CustomResourceDefinition: crd,
+		OutputPath:               outputPath,
+		Language:                 language,
+		APIVersion:               apiVersion,
+		Kind:                     kind,
+		Plural:                   plural,
+		Group:                    group,
+		Versions:                 versions,
+	}
+	return customResourceGenerator, nil
+}
+
+func (gen *CustomResourceGenerator) GetName() string {
+	return gen.Plural + "." + gen.Group
+}
+
+// GetVersionNames returns a slice of the versions supported by this CRD.
+func (gen *CustomResourceGenerator) GetVersionNames() []string {
+	versionNames := make([]string, 0, len(gen.Versions))
+	for versionName := range gen.Versions {
+		versionNames = append(versionNames, versionName)
+	}
+	return versionNames
+}
+
+func (gen *CustomResourceGenerator) GenerateCode() error {
+	objectTypeSpecs := GetObjectTypeSpecs(gen.Versions, gen.GetName(), gen.Kind)
+	PrettyPrint(objectTypeSpecs)
+
+	switch gen.Language {
+	case NodeJS:
+		types, err := nodejs.GenerateTypes(objectTypeSpecs)
+		if err != nil {
+			return fmt.Errorf("generate code: %v", err)
+		}
+		classes := gen.GenerateNodeJSClasses()
+
+		file, err := os.Create(gen.OutputPath)
+		if err != nil {
+			return fmt.Errorf("creating file at %s: %v", gen.OutputPath, err)
+		}
+		file.WriteString(types)
+		file.WriteString(classes)
+		defer file.Close()
```

Comment:
Nit: Move the `defer file.Close()` closer to the `os.Create` call:

```suggestion
		defer file.Close()
		file.WriteString(types)
		file.WriteString(classes)
```